File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.cpp
#pragma GCC optimize("O3", "unroll-loops");
static const int __ = [](){
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
class Solution {
   public:
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        sort(nums.begin(), nums.end());
        long long ans = 0;

        int n = nums.size();
        for (int i = 0; i < n; i++) {
            int l = lower - nums[i];
            int u = upper - nums[i];

            int llb = lower_bound(nums.begin()+i+1, nums.end(), l) - nums.begin();
            int rlb =
                lower_bound(nums.begin()+i+1, nums.end(), u + 1) - nums.begin();

            if(rlb > llb) ans += (rlb - llb);
        }

        return ans;
    }
};

File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.go
func lowerBound(x, l, r int, nums []int) int {

	for l <= r {
		mid := (l + r) >> 1
		if nums[mid] >= x {
			r = mid - 1
		} else {
			l = mid + 1
		}
	}

	return l
}
func countFairPairs(nums []int, lower int, upper int) int64 {
    ans := int64(0)
    sort.Ints(nums)
    for i, num := range nums {
        l := lower - num
        u := upper - num

        llb := lowerBound(l,i+1, len(nums)-1, nums)
        rlb := lowerBound(u+1, i+1, len(nums)-1, nums)

        if rlb > llb {
            ans += int64(rlb - llb)
        }
    }
    return ans
}
